src/core/server/integration_tests/saved_objects/migrations/zero_downtime/conversion_failures.test.ts

```diff
@@ -0,0 +1,198 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import Path from 'path';
+import fs from 'fs/promises';
+import { range } from 'lodash';
+import { createTestServers, type TestElasticsearchUtils } from '@kbn/core-test-helpers-kbn-server';
+import { SavedObjectsBulkCreateObject } from '@kbn/core-saved-objects-api-server';
+import '../jest_matchers';
+import { getKibanaMigratorTestKit } from '../kibana_migrator_test_kit';
+import { delay, parseLogFile } from '../test_utils';
+import { getBaseMigratorParams, getSampleAType, getSampleBType } from './base.fixtures';
+
+export const logFilePath = Path.join(__dirname, 'conversion_failures.test.log');
+
+describe('ZDT upgrades - encountering conversion failures', () => {
+  let esServer: TestElasticsearchUtils['es'];
+
+  const startElasticsearch = async () => {
+    const { startES } = createTestServers({
+      adjustTimeout: (t: number) => jest.setTimeout(t),
+      settings: {
+        es: {
+          license: 'basic',
+        },
+      },
+    });
+    return await startES();
+  };
+
+  beforeAll(async () => {
+    esServer = await startElasticsearch();
+  });
+
+  afterAll(async () => {
+    await esServer?.stop();
+    await delay(10);
+  });
+
+  beforeEach(async () => {
+    await fs.unlink(logFilePath).catch(() => {});
+  });
+
+  describe('when discardCorruptObjects is true', () => {
+    it('completes the migration and discard the documents', async () => {
+      const { runMigrations, savedObjectsRepository } = await prepareScenario({
+        discardCorruptObjects: true,
+      });
+
+      await runMigrations();
+
+      const records = await parseLogFile(logFilePath);
+      expect(records).toContainLogEntry('-> DONE');
+
+      const { saved_objects: sampleADocs } = await savedObjectsRepository.find({
+        type: 'sample_a',
+      });
+      const { saved_objects: sampleBDocs } = await savedObjectsRepository.find({
+        type: 'sample_b',
+      });
+
+      expect(sampleADocs).toHaveLength(0);
+      expect(sampleBDocs.map((doc) => doc.id).sort()).toEqual(['b-1', 'b-2', 'b-3', 'b-4']);
+    });
+  });
+
+  describe('when discardCorruptObjects is false', () => {
+    it('fails the migration with an explicit message and keep the documents', async () => {
+      const { runMigrations, savedObjectsRepository } = await prepareScenario({
+        discardCorruptObjects: false,
+      });
+
+      try {
+        await runMigrations();
+        expect('should have thrown').toEqual('but did not');
```

Comment:
nit: happy to go with your approach, however I believe there is a `fail('migration should have failed')` global.